DOC-5371 - two phase finality documentation and release notes (#1655)
corda/corda-docs-portal: content/en/platform/corda/4.11/community/api-flows.md

transaction fail to verify it, or the receiving flow (the finality handler) fails due to some other error, then
all parties will not have the up-to-date view of the ledger.

To recover from this scenario, the receiver’s finality handler is automatically sent to the `node-flow-hospital`. There, it is suspended and retried from its last checkpoint upon node restart, or according to other conditional retry rules. For more information, see [flow hospital runtime behavior]({{< relref "node-flow-hospital.md#run-time-behaviour" >}}).
This gives the node operator the opportunity to recover from the error. Until the issue is resolved, the node will continue to retry the flow
on each startup. Upon successful completion by the receiver’s finality flow, the ledger will become fully consistent.

{{< warning >}}
It is possible to forcibly terminate the error finality handler using the `killFlow` RPC. However, this risks an inconsistent view of the ledger.

{{< /warning >}}

#### Two Phase Finality

The Two Phase Finality protocol was introduced to improve resilience and recoverability.

> `FinalityFlow` will now:

Author: Done.

* Record the transaction locally without a notary signature.
* Broadcast the unnotarized transaction to other participants (for recording).
* Send the transaction to the chosen notary, and obtain a signature if the transaction is valid.
* Finalize the transaction locally with the notary signature.
* Broadcast the notary signature to other participants for finalization.

> `ReceiveFinalityFlow` will now:

Author: Done (both Ent and Community).

all parties will not have the up-to-date view of the ledger.

To recover from this scenario, the receiver’s finality handler is automatically sent to the `node-flow-hospital`. There, it is suspended and retried from its last checkpoint

> upon node restart, or according to other conditional retry rules - see [flow hospital runtime behavior]({{< relref "node-flow-hospital.md#run-time-behaviour" >}}).

Reviewer: ```suggestion
upon node restart, or according to other conditional retry rules. For more information, see [flow hospital runtime behavior]({{< relref "node-flow-hospital.md#run-time-behaviour" >}}).
```

{{< /warning >}}

#### Two Phase Finality

> From Corda 4.11 on, the finality protocol has changed to improve resilience and recoverability.

Reviewer: This reads like a release note?

Author: Rewrote these to omit "From Corda 4.11 on".

> `FinalityFlow` will now:

Reviewer: Again, seems like a release note?

Implementations of `FlowExternalAsyncOperation` must return a `CompletableFuture`. The developer decides how to create this future.
The best practice is to use `CompletableFuture.supplyAsync` and supply an executor to run the future. You can use other libraries to

> generate futures, as long as a `CompletableFuture` is returned out of `FlowExternalAsyncOperation`. You can see an example of creating a future using [Guava’s ListenableFuture](#api-flows-guava-future-conversion) Below.

Author: Done (both Ent and Community)

Implementations of `FlowExternalAsyncOperation` must return a `CompletableFuture`. The developer decides how to create this future.
The best practice is to use `CompletableFuture.supplyAsync` and supply an executor to run the future. You can use other libraries to

> generate futures, as long as a `CompletableFuture` is returned out of `FlowExternalAsyncOperation`. You can see an example of creating a future using [Guava’s ListenableFuture](#api-flows-guava-future-conversion) Below.

Reviewer: ```suggestion
generate futures, as long as a `CompletableFuture` is returned out of `FlowExternalAsyncOperation`. An example of creating a future using [Guava’s ListenableFuture](#api-flows-guava-future-conversion) is shown below.
```

Can we have a cross-ref or something rather than using "below"?

Author: Added anchors to both Ent and Community so link works.

Implementations of `FlowExternalAsyncOperation` must return a `CompletableFuture`. The developer decides how to create this future.
The best practice is to use `CompletableFuture.supplyAsync` and supply an executor to run the future. You can use other libraries to
generate futures, as long as a `CompletableFuture` is returned out of `FlowExternalAsyncOperation`. You can see an example of creating a future using [Guava’s ListenableFuture](#api-flows-guava-future-conversion) below.

> generate futures, as long as a `CompletableFuture` is returned out of `FlowExternalAsyncOperation`. You can see an example of creating a future using [Guava’s ListenableFuture](#api-flows-guava-future-conversion) below.

Reviewer: Seems to be a duplicate line?

You shouldn't use in-memory data structures to handle deduplication as their state will not survive node restarts.

{{< /warning >}}

> <a name="api-flows-guava-future-conversion"></a>

Author: Inserted an anchor here for that [Guava’s ListenableFuture](#api-flows-guava-future-conversion) link (as there was no such heading).
